### PR TITLE
fix: getYaw

### LINKED ISSUE #62: Possible issue in Pose3D.getYaw
I had an issue in the emc_simulator in which I was trying to extract the yaw from a Pose3D. I replicated the issue in the following code snippet
```
#include <geolib/datatypes.h>

int main()
{
   double x,y,z,roll,pitch,yaw;
   x = 0;
   y = 0;
   z = 0;  
   
   roll  = 0;
   pitch = 0;
   yaw   = 1;
   
   geo::Pose3D pose = geo::Pose3D(x,y,z,roll,pitch,yaw);
   std::cout<<pose.getYaw()<<std::endl;  
}
```
I expect the output of the program to be 1, since I set the yaw of the Pose3D to 1, however, the output that is printed to the terminal is equal to 0. When I set the roll to 1, and yaw to zero, the result that is printed **is** equal to 1.

### PR DESCRIPTION
**Pull request by KdVos**
Fix for #62

### CODE REVIEW

File: include/geolib/math_types.h [hunk 1]
```diff
@@ -747,9 +747,22 @@ class Transform3T {
...
+            // At singularity roll is set to zero, yaw is equal to sum of both
+            return yaw;
+        }
+        else
+        {
+            roll = atan2(R.zy, R.zz);
+            yaw  = atan2(R.yx, R.xx);
+            return yaw;
+        }
     }
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]: Better to make a function getRPY. Which is then called by getYaw. As you care now also calculating the roll, which is not used. But both the roll and pitch can be usefull.

File: include/geolib/math_types.h [hunk 2]
```diff
@@ -747,9 +747,22 @@ class Transform3T {
...
-        QuaternionT<T> q;
-        R.getRotation(q);
-        return atan2(2.0*(q.y*q.z + q.w*q.x), q.w*q.w - q.x*q.x - q.y*q.y + q.z*q.z);
+        double roll, pitch, yaw;        
+        double epsilon = 1e-12;
+        pitch = atan2(-R.zx, sqrt( R.zy*R.zy + R.zz*R.zz));
+        if (std::fabs(pitch - M_PI/2) < epsilon) // detect singularity
+        {
+            yaw = atan2(-R.xy, R.yy);
+            roll= 0;
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]:
```suggestion
            roll = 0;
```

File: include/geolib/math_types.h [hunk 3]
```diff
@@ -747,9 +747,22 @@ class Transform3T {
...
+            // At singularity roll is set to zero, yaw is equal to sum of both
+            return yaw;
+        }
+        else
+        {
+            roll = atan2(R.zy, R.zz);
+            yaw  = atan2(R.yx, R.xx);
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]:
```suggestion
            yaw = atan2(R.yx, R.xx);
```

File: include/geolib/math_types.h [hunk 4]
```diff
@@ -747,9 +747,22 @@ class Transform3T {
     }
 
     double getYaw()  {
-        QuaternionT<T> q;
-        R.getRotation(q);
-        return atan2(2.0*(q.y*q.z + q.w*q.x), q.w*q.w - q.x*q.x - q.y*q.y + q.z*q.z);
+        double roll, pitch, yaw;        
+        double epsilon = 1e-12;
+        pitch = atan2(-R.zx, sqrt( R.zy*R.zy + R.zz*R.zz));
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]:
```suggestion
        pitch = atan2(-R.zx, sqrt(R.zy*R.zy + R.zz*R.zz));
```

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]

File: include/geolib/math_types.h [hunk 5]
```diff
@@ -745,11 +745,30 @@ class Transform3T {
     void setRPY(T roll, T pitch, T yaw)  {
         R.setRPY(roll, pitch, yaw);
     }
+    
+    void getRPY(T &roll, T &pitch, T& yaw){
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]:
```suggestion
    void getRPY(T& roll, T& pitch, T& yaw) const {
```

File: include/geolib/math_types.h [hunk 6]
```diff
@@ -745,11 +745,30 @@ class Transform3T {
...
+        double epsilon = 1e-12;
+        pitch = atan2(-R.zx, sqrt(R.zy*R.zy + R.zz*R.zz));
+        if (std::fabs(pitch - M_PI/2) < epsilon) // detect singularity
+        {
+            yaw = atan2(-R.xy, R.yy);
+            roll = 0;
+            // At singularity roll is set to zero, yaw is equal to sum of both
+            return;
+        }
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]:
```suggestion
        {
            // At singularity roll is set to zero, yaw is equal to sum of both
            yaw = atan2(-R.xy, R.yy);
            roll = 0;
        }
```

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]

File: include/geolib/math_types.h [hunk 7]
```diff
@@ -745,11 +745,30 @@ class Transform3T {
...
+    }
 
     double getYaw()  {
-        QuaternionT<T> q;
-        R.getRotation(q);
-        return atan2(2.0*(q.y*q.z + q.w*q.x), q.w*q.w - q.x*q.x - q.y*q.y + q.z*q.z);
+    	T roll, pitch, yaw;
+    	getRPY(roll, pitch, yaw);
+    	return yaw;
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]: This is also a const function.

File: include/geolib/math_types.h (hunk 5 above)
Comment (MatthijsBurgh) [reply include/geolib/math_types.h]: I have really no clue how a Transform of an int or unsigned should look. But I don't  think it's RPY will be an int/unsigned. Also this conflicts with `getYaw`, which always returns a double.

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]

File: include/geolib/math_types.h (hunk 5 above)
Comment (KdVos) [reply include/geolib/math_types.h]: True, however, I wanted to keep the input/output consistent with ` void setRPY(T roll, T pitch, T yaw) ` above

Comment (MatthijsBurgh) [reply include/geolib/math_types.h]: Lets change both.

Comment (KdVos) [reply include/geolib/math_types.h]: done

File: include/geolib/math_types.h [hunk 8]
```diff
@@ -742,14 +742,33 @@ class Transform3T {
         return Transform3T(inv, inv * -t);
     }
 
-    void setRPY(T roll, T pitch, T yaw)  {
+    void setRPY(double roll, double pitch, double yaw) {
         R.setRPY(roll, pitch, yaw);
     }
+    
+    void getRPY(double& roll, double& pitch, double& yaw) const{
```
Comment (MatthijsBurgh) [inline include/geolib/math_types.h]:
```suggestion
    void getRPY(double& roll, double& pitch, double& yaw) const {
```

**Review by MatthijsBurgh — APPROVED** [review]

**Comment by MatthijsBurgh** [comment]
@KdVos did we consider https://github.com/ros/geometry/blob/d51887c976a45ba01db9c61361e94076d3bcf3f1/tf/include/tf/LinearMath/Matrix3x3.h#L294?
It looks much more extended than our code.

**Pull request merged**